tool/python/singa/layer.py

```diff
@@ -0,0 +1,431 @@
+        fields = {'pool' : PoolingProto().AVG,
+                  'kernel' : pool_size[0],
+                  'stride' : stride,
+                  'pad' : 0 if ignore_border else 1}
+        setval(self.layer.pooling_conf, **fields)
+
+class LRN2D(Layer):
+
+    def __init__(self, size=0, **kwargs):
+        '''
+        required
+          size = (int)  // local size
+        '''
+
+        super(LRN2D, self).__init__(name=generate_name('norm'), type=kLRN)
+        # required
+        assert size != 0, 'local size should be set'
+        self.layer.lrn_conf.local_size = size
+        init_values = get_init_values('lrn2d', **kwargs)
+        setval(self.layer.lrn_conf, **init_values)
+
+
+class Activation(Layer):
+
+    def __init__(self, activation='stanh', topk=1):
+        '''
+        required
+          activation = (string)
+        optional
+          topk       = (int)  // the number of results
+        '''
+
+        self.name = activation
+        if activation == 'tanh': activation = 'stanh' # <-- better way to set?
+        self.layer_type = enumLayerType(activation)
+        super(Activation, self).__init__(name=generate_name(self.name),
+                                         type=self.layer_type)
+        if activation == 'softmaxloss':
+            self.layer.softmaxloss_conf.topk = topk
+
+class Dropout(Layer):
+
+    def __init__(self, ratio=0.5):
+        '''
+        required
+          ratio = (float) // ratio of drop out nodes
+        '''
+
+        self.name = 'dropout'
+        self.layer_type = enumLayerType(self.name)
+        super(Dropout, self).__init__(name=generate_name(self.name),
+                                      type=self.layer_type)
+        self.layer.dropout_conf.dropout_ratio = ratio
+
+
+class RGB(Layer):
+
+    def __init__(self, meanfile=None, **kwargs):
+        '''
+        required
+          meanfile = (string) // path to meanfile (depreciated)
+        '''
+
+        assert meanfile != None, 'meanfile should be specified'
+        self.name = 'rgb'
+        self.layer_type = kRGBImage
+        super(RGB, self).__init__(name=generate_name(self.name),
+                                  type=self.layer_type)
+        self.layer.rgbimage_conf.meanfile = meanfile
+
+class Dense(Layer):
+
+    def __init__(self, output_dim=0, activation=None,
+                 init=None, w_param=None, b_param=None, input_dim=None,
+                 **kwargs):
+        '''
+        required
+          output_dim = (int)
+        optional
+          activation = (string)
+          init       = (string)     // 'unirom', 'gaussian', 'constant'
+          w_param    = (Parameter)  // Parameter object for weight
+          b_param    = (Parameter)  // Parameter object for bias
+          **kwargs
+            w_lr = (float) // learning rate for w
+            w_wd = (float) // weight decay for w
+            b_lr = (float) // learning rate for b
+            b_wd = (float) // weight decay for b
```

Comment:
Thank you!
Just be careful about xx_lr and xx_wd, which are different to lr and wd.
the first set is multiplier, which scales the lr/wd.

On Fri, Jan 1, 2016 at 6:34 PM, chonho <notifications@github.com> wrote:

> In tool/python/singa/layer.py
> <https://github.com/apache/incubator-singa/pull/106#discussion_r48677121>:
>
> > +    def __init__(self, output_dim=0, activation=None,
> > +                 init=None, w_param=None, b_param=None, input_dim=None,
> > +                 **kwargs):
> > +        '''
> > +        required
> > +          output_dim = (int)
> > +        optional
> > +          activation = (string)
> > +          init       = (string)     // 'unirom', 'gaussian', 'constant'
> > +          w_param    = (Parameter)  // Parameter object for weight
> > +          b_param    = (Parameter)  // Parameter object for bias
> > +          **kwargs
> > +            w_lr = (float) // learning rate for w
> > +            w_wd = (float) // weight decay for w
> > +            b_lr = (float) // learning rate for b
> > +            b_wd = (float) // weight decay for b
>
> I will update it soon
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-singa/pull/106/files#r48677121>.
>